File: Python/sql_generator.py
def quote_value(value: str) -> str:
    value = value.strip()
    if value.lower() == "null":
        return "NULL"

    try:
        float(value)
        return value
    except ValueError:
        escaped = value.replace("'", "''")
        return f"'{escaped}'"


def build_select() -> str:
    table = input("Table name: ").strip()
    columns = input("Columns (comma-separated, or *): ").strip() or "*"
    where = input("WHERE condition (optional): ").strip()

    sql = f"SELECT {columns} FROM {table}"
    if where:
        sql += f" WHERE {where}"
    return sql + ";"


def build_insert() -> str:
    table = input("Table name: ").strip()
    columns_raw = input("Columns (comma-separated): ").strip()
    values_raw = input("Values (comma-separated): ").strip()

    columns = [c.strip() for c in columns_raw.split(",") if c.strip()]
    values = [v.strip() for v in values_raw.split(",")]

    if len(columns) != len(values):
        raise ValueError("Columns count must match values count.")

    formatted_values = ", ".join(quote_value(v) for v in values)
    columns_part = ", ".join(columns)

    return f"INSERT INTO {table} ({columns_part}) VALUES ({formatted_values});"


def build_update() -> str:
    table = input("Table name: ").strip()
    assignments_raw = input("Set values (example: name=John, age=25): ").strip()
    where = input("WHERE condition (required for safety): ").strip()

    if not where:
        raise ValueError("WHERE condition is required for UPDATE.")

    assignments = []
    for part in assignments_raw.split(","):
        if "=" not in part:
            raise ValueError(f"Invalid assignment: {part}")
        col, val = part.split("=", 1)
        assignments.append(f"{col.strip()}={quote_value(val.strip())}")

    set_part = ", ".join(assignments)
    return f"UPDATE {table} SET {set_part} WHERE {where};"


def build_delete() -> str:
    table = input("Table name: ").strip()
    where = input("WHERE condition (required for safety): ").strip()

    if not where:
        raise ValueError("WHERE condition is required for DELETE.")

    return f"DELETE FROM {table} WHERE {where};"


def main() -> None:
    print("Simple SQL Generator")
    print("1) SELECT")
    print("2) INSERT")
    print("3) UPDATE")
    print("4) DELETE")

    choice = input("Choose query type (1-4): ").strip()

    try:
        if choice == "1":
            sql = build_select()
        elif choice == "2":
            sql = build_insert()
        elif choice == "3":
            sql = build_update()
        elif choice == "4":
            sql = build_delete()
        else:
            print("Invalid choice.")
            return

        print("\nGenerated SQL:")
        print(sql)

    except ValueError as error:
        print(f"Error: {error}")


if __name__ == "__main__":
    main()

File: Typescript/sql_generator.ts
const readline = require('node:readline/promises');
const { stdin, stdout } = require('node:process');

const quoteValue = (value) => {
  const trimmed = value.trim();

  if (trimmed.toLowerCase() === 'null') {
    return 'NULL';
  }

  if (!Number.isNaN(Number(trimmed))) {
    return trimmed;
  }

  return `'${trimmed.replaceAll("'", "''")}'`;
};

const buildSelect = async (rl) => {
  const table = (await rl.question('Table name: ')).trim();
  const columns = (await rl.question('Columns (comma-separated, or *): ')).trim() || '*';
  const where = (await rl.question('WHERE condition (optional): ')).trim();

  let sql = `SELECT ${columns} FROM ${table}`;
  if (where) {
    sql += ` WHERE ${where}`;
  }

  return `${sql};`;
};

const buildInsert = async (rl) => {
  const table = (await rl.question('Table name: ')).trim();
  const columnsRaw = (await rl.question('Columns (comma-separated): ')).trim();
  const valuesRaw = (await rl.question('Values (comma-separated): ')).trim();

  const columns = columnsRaw
    .split(',')
    .map((item) => item.trim())
    .filter(Boolean);
  const values = valuesRaw.split(',').map((item) => item.trim());

  if (columns.length !== values.length) {
    throw new Error('Columns count must match values count.');
  }

  const formattedValues = values.map(quoteValue).join(', ');
  return `INSERT INTO ${table} (${columns.join(', ')}) VALUES (${formattedValues});`;
};

const buildUpdate = async (rl) => {
  const table = (await rl.question('Table name: ')).trim();
  const assignmentsRaw = (await rl.question('Set values (example: name=John, age=25): ')).trim();
  const where = (await rl.question('WHERE condition (required for safety): ')).trim();

  if (!where) {
    throw new Error('WHERE condition is required for UPDATE.');
  }

  const assignments = assignmentsRaw.split(',').map((part) => {
    if (!part.includes('=')) {
      throw new Error(`Invalid assignment: ${part}`);
    }

    const [column, value] = part.split('=');
    return `${column.trim()}=${quoteValue(value.trim())}`;
  });

  return `UPDATE ${table} SET ${assignments.join(', ')} WHERE ${where};`;
};

const buildDelete = async (rl) => {
  const table = (await rl.question('Table name: ')).trim();
  const where = (await rl.question('WHERE condition (required for safety): ')).trim();

  if (!where) {
    throw new Error('WHERE condition is required for DELETE.');
  }

  return `DELETE FROM ${table} WHERE ${where};`;
};

const main = async () => {
  const rl = readline.createInterface({ input: stdin, output: stdout });

  try {
    console.log('Simple SQL Generator');
    console.log('1) SELECT');
    console.log('2) INSERT');
    console.log('3) UPDATE');
    console.log('4) DELETE');

    const choice = (await rl.question('Choose query type (1-4): ')).trim();

    let sql;
    if (choice === '1') {
      sql = await buildSelect(rl);
    } else if (choice === '2') {
      sql = await buildInsert(rl);
    } else if (choice === '3') {
      sql = await buildUpdate(rl);
    } else if (choice === '4') {
      sql = await buildDelete(rl);
    } else {
      console.log('Invalid choice.');
      return;
    }

    console.log('\nGenerated SQL:');
    console.log(sql);
  } catch (error) {
    console.log(`Error: ${error.message}`);
  } finally {
    rl.close();
  }
};

main();
